### PR TITLE
[WIP] Refactor policy and endpoint packages

### PR DESCRIPTION
**Pull request by aanm**
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/8526)

### CODE REVIEW

File: pkg/endpoint/regeneration/owner.go [hunk 1]
```diff
@@ -103,3 +106,149 @@ type EndpointUpdater interface {
...
+type SelectorCache interface {
+	RemoveSelectors(CachedSelectorSlice, CachedSelectionUser)
+	RemoveSelector(CachedSelector, CachedSelectionUser)
+	ChangeUser(CachedSelector, CachedSelectionUser, CachedSelectionUser)
+	AddIdentitySelector(CachedSelectionUser, api.EndpointSelector) (CachedSelector, bool)
+	AddFQDNSelector(CachedSelectionUser, api.FQDNSelector) (CachedSelector, bool)
+}
+
+type PolicyL4Filter interface {
+	L7ParserType() string
```
Comment (aanm) [inline pkg/endpoint/regeneration/owner.go:134]: This should return `policy.L7ParserType` but then we will import policy into this package again. Any better ideais @joestringer?

File: pkg/endpoint/bpf.go [hunk 2]
```diff
@@ -214,13 +214,15 @@ func (e *Endpoint) addNewRedirectsFromMap(owner regeneration.Owner, m policy.L4P
 			keysFromFilter := l4.ToKeys(direction)
```
Comment (aanm) [inline pkg/endpoint/bpf.go:214]: The `l4.ToKeys` could return `PolicyKeys` directly so you wouldn't need to cast it all the time on line 217.

File: pkg/endpoint/regeneration/owner.go [hunk 3]
```diff
@@ -103,3 +106,149 @@ type EndpointUpdater interface {
...
+
+	// String returns the string representation of this selector.
+	// Used as a map key.
+	String() string
+}
+
+// CachedSelectorSlice is a slice of CachedSelectors that can be sorted.
+type CachedSelectorSlice []CachedSelector
+
+func (s CachedSelectorSlice) Len() int      { return len(s) }
```
Comment (aanm) [inline pkg/endpoint/regeneration/owner.go:203]: These methods can be replaced with https://golang.org/pkg/sort/#Slice

File: pkg/endpoint/regeneration/owner.go [hunk 4]
```diff
@@ -103,3 +106,149 @@ type EndpointUpdater interface {
...
+	for _, selector := range s {
+		if selector.IsWildcard() {
+			return true
+		}
+	}
+	return false
+}
+
+// Insert in a sorted order? Returns true if inserted, false if cs was already in
+func (s *CachedSelectorSlice) Insert(cs CachedSelector) bool {
```
Comment (aanm) [inline pkg/endpoint/regeneration/owner.go:223]: Why does this implementation exist here?

**Comment by ianvernon** [comment]
I know this is WIP, but what is the need for this?

**Comment by fristonio** [comment]
Hey @ianvernon 
This is with reference to [this](https://github.com/cilium/cilium/issues/8364) issue. The Pull Request is just to see the diffs for bringing the change as requested by aanm. I will update the title and description of the Pull Request in a while.

**Pull request closed without merging**